### PR TITLE
[Metricbeat] Run Go integration tests sequentially

### PR DESCRIPTION
**Pull request by mtojek**
Ok, now this PR needs a couple of notes.

Go tests run in parallel which means that too many containers are spawned in the same time. This leads to read timeouts, build failures, flakiness, etc.
The PR adjusts test execution to verify only one module in the same time. 
Also, updated app-search module and added small waiting routine for Elasticsearch endpoint.
Verbose mode has been enabled for Go tests to allow go-junit-report properly parse and report stats (missing PASS entries).

~~This PR adds a retry loop for appsearch.~~

~~While investigating I managed to replicate some flakiness:~~

```
---- appsearch_1
Found java executable in PATH
Java version: 1.8.0_222

Running Elastic App Search in a single-process mode (process = app-server)!

WARNING: Failure to run all of the components of Elastic App Search will result in a broken service!
         This is an unsupported mode! The single-process mode should only be used in advanced
         situations when all of App Search components are deployed and scaled separately.

scripting container class loader urls: [file:/tmp/jruby1382536578408950512extract/lib/jruby-core-9.2.5.0-complete.jar, file:/tmp/jruby1382536578408950512extract/lib/jruby-rack-1.1.21.jar, file:/tmp/jruby1382536578408950512extract/lib/jruby-stdlib-9.2.5.0.jar]
setting GEM_HOME to /tmp/jruby1382536578408950512extract/gems
... and BUNDLE_GEMFILE to /tmp/jruby1382536578408950512extract/Gemfile
loading resource: /tmp/jruby1382536578408950512extract/./META-INF/rails.rb
invoking /tmp/jruby1382536578408950512extract/./META-INF/rails.rb with: [runner, LocoTogo.start_app_server!]
[2019-12-16T08:40:26.917+00:00][15][2002][app-server][ERROR]:
--------------------------------------------------------------------------------

Elasticsearch cluster must be licensed. OSS versions of Elasticsearch do not contain a supported license. Please download and run an Elasticsearch binary from https://elastic.co/downloads/elasticsearch to acquire a free, Basic license.

--------------------------------------------------------------------------------


ERROR: org.jruby.exceptions.SystemExit: (SystemExit) exit
org.jruby.exceptions.SystemExit: (SystemExit) exit
	at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:741)
	at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:704)
	at RUBY.fatal_error(/tmp/jruby1382536578408950512extract/shared_togo/lib/shared_togo.class:694)
	at RUBY.check_elasticsearch_license!(/tmp/jruby1382536578408950512extract/loco_togo/lib/loco_togo.class:64)
	at RUBY.configure!(/tmp/jruby1382536578408950512extract/shared_togo/lib/shared_togo.class:177)
	at RUBY.configure!(/tmp/jruby1382536578408950512extract/shared_togo/lib/shared_togo.class:31)
	at RUBY.<main>(/tmp/jruby1382536578408950512extract/config/application.class:24)
	at org.jruby.RubyKernel.load(org/jruby/RubyKernel.java:1007)
	at RUBY.<main>(/tmp/jruby1382536578408950512extract/config/application.rb:1)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:983)
	at RUBY.<main>(/tmp/jruby1382536578408950512extract/config/application.rb:1)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:983)
	at RUBY.(root)(/tmp/jruby1382536578408950512extract/gems/gems/railties-4.2.11.1/lib/rails/commands/runner.rb:1)
	at RUBY.<main>(/tmp/jruby1382536578408950512extract/gems/gems/railties-4.2.11.1/lib/rails/commands/runner.rb:51)
	at RUBY.(root)(/tmp/jruby1382536578408950512extract/gems/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:1)
	at RUBY.<main>(/tmp/jruby1382536578408950512extract/gems/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:123)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:983)
	at tmp.jruby1382536578408950512extract.$_dot_.META_minus_INF.rails.<main>(/tmp/jruby1382536578408950512extract/./META-INF/rails.rb:7)
```

### CODE REVIEW

**Comment by mtojek** [comment]
jenkins, test this please

**Comment by mtojek** [comment]
jenkins, test this please

**Comment by ChrsMark** [comment]
This is really nice! 
To my mind it would be nice to have info of Docker build phase. Currently, for instance in x-pack testsuite we have:
```
>> Building docker images
Building beat
Building elasticsearch
Building kibana
Creating network "metricbeat_8_0_0_379d23b275-snapshot_default" with the default driver
2019/12/16 18:47:54 Found Elastic Beats dir at /go/src/github.com/elastic/beats
>> go test: Integration-activemq Testing
```
It would be nice to see info like:
```
Building coredns
```
Also what if a build of the container of an under-test service, like activemq, fails. Will we learn about this?

**Comment by mtojek** [comment]
@ChrsMark  To my understanding it will end-up in timeout (unhealthy service status, even if it's failing all the time). I'm not sure if it should be covered in the scope of this task, because it's rather an improvement in "compose" part, but I can work on the improvement of that one later on!

**Comment by mtojek** [comment]
Here is a sample output: https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-linux/4700/beat=x-pack%2Fmetricbeat,label=linux-immutable/console

File: x-pack/metricbeat/magefile.go [hunk 1]
```diff
@@ -154,7 +156,25 @@ func GoUnitTest(ctx context.Context) error {
 // Use RACE_DETECTOR=true to enable the race detector.
 func GoIntegTest(ctx context.Context) error {
 	return devtools.RunIntegTest("goIntegTest", func() error {
-		return devtools.GoTest(ctx, devtools.DefaultGoTestIntegrationArgs())
+		modulesFileInfo, err := ioutil.ReadDir("./module")
```
Comment (exekias) [inline x-pack/metricbeat/magefile.go]: should we do this change also for non-xpack metricbeat?

Comment (mtojek) [reply x-pack/metricbeat/magefile.go]: I believe that at some point we may come to the same limit. Also, we can face any interferences between tests right now.

Comment (exekias) [reply x-pack/metricbeat/magefile.go]: also having both tested in the same way should reduce complexity. +1 to doing it now

Comment (mtojek) [reply x-pack/metricbeat/magefile.go]: Change added.

**Comment by mtojek** [comment]
jenkins, test this please

File: x-pack/metricbeat/magefile.go [hunk 2]
```diff
@@ -154,7 +156,25 @@ func GoUnitTest(ctx context.Context) error {
...
+		if err != nil {
+			return err
+		}
+
+		var failed bool
+		for _, fi := range modulesFileInfo {
+			if !fi.IsDir() {
+				continue
+			}
+			err := devtools.GoTest(ctx, devtools.GoTestIntegrationArgsForModule(fi.Name()))
```
Comment (exekias) [inline x-pack/metricbeat/magefile.go]: I'm wondering now, this will actually start the containers sequentially, but is it stopping them when the test ends?

Comment (mtojek) [reply x-pack/metricbeat/magefile.go]: Nope, good catch. There wasn't such mechanism at all. I will add it here.

Comment (mtojek) [reply x-pack/metricbeat/magefile.go]: After discussing it offline, we'll come back to this feature later.

**Comment by mtojek** [comment]
jenkins, test this please

**Comment by jsoriano** [comment]
> It would be nice to see info like:
> 
> ```
> Building coredns
> ```
> 
> Also what if a build of the container of an under-test service, like activemq, fails. Will we learn about this?

Don't worry so much about image building, we are moving the prebuild of images to a separate job.

File: libbeat/scripts/Makefile [hunk 3]
```diff
@@ -181,9 +181,9 @@ unit: ## @testing Runs the unit tests without coverage reports.
 
 .PHONY: integration-tests
 integration-tests: ## @testing Run integration tests. Unit tests are run as part of the integration tests.
-integration-tests: prepare-tests
+integration-tests: prepare-tests mage
 	rm -f docker-compose.yml.lock
-	$(COVERAGE_TOOL) -tags=integration $(RACE) -coverprofile=${COVERAGE_DIR}/integration.cov ${GOPACKAGES}
+	mage GoTestIntegration || $(COVERAGE_TOOL) -tags=integration $(RACE) -coverprofile=${COVERAGE_DIR}/integration.cov ${GOPACKAGES}
```
[the remark this answers is not in the document]
Comment (mtojek) [reply libbeat/scripts/Makefile]: Actually I tried it and it failed for auditbeat with a strange notice informing about skipping tests due to lack of docker...

Maybe it did sth wrong, will try again.

**Comment by mtojek** [comment]
jenkins, test this please

File: dev-tools/mage/target/integtest/integtest.go [hunk 4]
```diff
@@ -62,12 +62,7 @@ func IntegTest() {
 // Use TEST_COVERAGE=true to enable code coverage profiling.
 // Use RACE_DETECTOR=true to enable the race detector.
 func GoIntegTest(ctx context.Context) error {
-	if !devtools.IsInIntegTestEnv() {
-		mg.SerialDeps(goTestDeps...)
-	}
-	return devtools.RunIntegTest("goIntegTest", func() error {
```
Comment (andrewkroh) [inline dev-tools/mage/target/integtest/integtest.go]: Why the change to not run `go test -tags integration` inside of a container? Won't this break some tests expecting containerized services?

Comment (mtojek) [reply dev-tools/mage/target/integtest/integtest.go]: So far I hit the other problem (for auditbeat):

```
22:28:21 mage goIntegTest
22:28:37 2019/12/17 22:28:37 Found Elastic Beats dir at /go/src/github.com/elastic/beats
22:28:37 >> goIntegTest: Skipping because docker is not available
```

**Comment by mtojek** [comment]
In general, Jenkins CI (beats-ci) passed.

Error reported Travis CI due to "The job exceeded the maximum log length, and has been terminated." - will fix it in a moment

**Comment by mtojek** [comment]
Issue with noisy logs fixed.

@jsoriano could you please take a look at recent changes?

File: libbeat/scripts/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply libbeat/scripts/Makefile]: Ok, managed to fix it.

**Comment by mtojek** [comment]
jenkins, test this please

File: libbeat/publisher/queue/spool/spool_test.go [hunk 5]
```diff
@@ -44,9 +45,12 @@ type testLogger struct {
 	t *testing.T
 }
 
+type silentLogger struct{}
+
 func init() {
 	flag.Int64Var(&seed, "seed", time.Now().UnixNano(), "test random seed")
 	flag.BoolVar(&debug, "noisy", false, "print test logs to console")
+	flag.BoolVar(&verboseLogger, "verboseLogger", false, "use verbose logger")
```
Comment (jsoriano) [inline libbeat/publisher/queue/spool/spool_test.go]: Is it unused? And shouldn't it be enough with the `noisy` flag?

File: dev-tools/mage/gotest.go [hunk 6]
```diff
@@ -94,6 +118,33 @@ func DefaultTestBinaryArgs() TestBinaryArgs {
 	}
 }
 
+// GoTestIntegrationForModule executes the Go integration tests.
```
Comment (jsoriano) [inline dev-tools/mage/gotest.go]: Maybe we should mention that tests here are run sequentially, and maybe explain the motivation of it.

File: dev-tools/mage/gotest.go [hunk 7]
```diff
@@ -94,6 +118,33 @@ func DefaultTestBinaryArgs() TestBinaryArgs {
...
+// Use TEST_COVERAGE=true to enable code coverage profiling.
+// Use RACE_DETECTOR=true to enable the race detector.
+func GoTestIntegrationForModule(ctx context.Context) error {
+	return RunIntegTest("goIntegTest", func() error {
+		modulesFileInfo, err := ioutil.ReadDir("./module")
```
Comment (jsoriano) [inline dev-tools/mage/gotest.go]: Not all beats have their "modules" under `module`, maybe the modules path should be parameterized. But we can change this in a future change if we find other beats needing this functionality.

File: auditbeat/magefile.go [hunk 8]
```diff
@@ -169,29 +164,8 @@ func GoUnitTest(ctx context.Context) error {
 	return devtools.GoTest(ctx, devtools.DefaultGoTestUnitArgs())
 }
 
-// GoIntegTest executes the Go integration tests.
-// Use TEST_COVERAGE=true to enable code coverage profiling.
-// Use RACE_DETECTOR=true to enable the race detector.
-func GoIntegTest(ctx context.Context) error {
-	mg.Deps(Fields)
```
Comment (jsoriano) [inline auditbeat/magefile.go]: Some integration tests we are removing in this PR depended on `Fields`, are we keeping these dependencies or they were not needed?

File: libbeat/publisher/queue/spool/spool_test.go (hunk 5 above)
Comment (mtojek) [reply libbeat/publisher/queue/spool/spool_test.go]: The flag `noisy` (debug) means that you want to log more data than just single log line:

```
func (l *testLogger) reportf(level string, str string, vs []interface{}) {
	str = level + ": " + str
	l.t.Logf(str, vs...)
	if debug {
		fmt.Printf(str, vs...)
	}
}
```
Leaving it as is prints thousands of log lines and exceeds Travis limit, hence the silent logger.

File: dev-tools/mage/gotest.go (hunk 7 above)
Comment (mtojek) [reply dev-tools/mage/gotest.go]: I think we may come up to this in the future, right now only Metricbeat (and x-pack Metricbeat) uses this feature.

File: auditbeat/magefile.go (hunk 8 above)
Comment (mtojek) [reply auditbeat/magefile.go]: Hmm, you need to give me more references. I can confirm that all tests passed in the Beats CI.

File: dev-tools/mage/gotest.go (hunk 6 above)
Comment (mtojek) [reply dev-tools/mage/gotest.go]: Good point, done.

**Comment by mtojek** [comment]
jenkins, test this please

File: x-pack/metricbeat/magefile.go [hunk 9]
```diff
@@ -154,7 +156,25 @@ func GoUnitTest(ctx context.Context) error {
 // Use RACE_DETECTOR=true to enable the race detector.
 func GoIntegTest(ctx context.Context) error {
 	return devtools.RunIntegTest("goIntegTest", func() error {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply x-pack/metricbeat/magefile.go]: Fixed

File: libbeat/publisher/queue/spool/spool_test.go (hunk 5 above)
Comment (mtojek) [reply libbeat/publisher/queue/spool/spool_test.go]: You're right, I will use a single flag for debug/noisy logger.

**Comment by mtojek** [comment]
jenkins, test this please

**Comment by mtojek** [comment]
Beats CI errors irrelevant `(test.TestSecurity/security-windows*)`.
Travic CI passed.

I will proceed with merging.
